subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +592,460 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
-	}
-
-	LOG_DBG("start fragmenting");
-	/*
-	 * Send the fragments. For the last one simply use the original
-	 * buffer (which works since we've used net_buf_pull on it).
+	/* TODO: This function should be overridable by the application: they
+	 * should be able to provide their own heuristic.
 	 */
-	flags = FRAG_START;
-	if (tx_data(buf)->is_cont) {
-		flags = FRAG_CONT;
+	if (!conn->has_data(conn)) {
+		LOG_DBG("No more data for %p", conn);
+		return true;
 	}
 
-	while (buf->len > conn_mtu(conn)) {
-		frag = create_frag(conn, buf);
-		if (!frag) {
-			return -ENOMEM;
-		}
-
-		err = send_frag(conn, buf, frag, flags);
-		if (err) {
-			LOG_DBG("%p failed, mark as existing frag", buf);
-			tx_data(buf)->is_cont = flags != FRAG_START;
-			net_buf_unref(frag);
-			return err;
-		}
-
-		flags = FRAG_CONT;
+	/* Queue only 3 buffers per-conn for now */
+	if (atomic_get(&conn->in_ll) < 3) {
+		/* The goal of this heuristic is to allow the link-layer to
+		 * extend an ACL connection event as long as the application
+		 * layer can provide data.
+		 *
+		 * Here we chose three buffers, as some LLs need two enqueued
+		 * packets to be able to set the more-data bit, and one more
+		 * buffer to allow refilling by the app while one of them is
+		 * being sent over-the-air.
+		 */
+		return false;
 	}
 
-	LOG_DBG("last frag");
-	tx_data(buf)->is_cont = true;
-	return send_frag(conn, buf, NULL, FRAG_END);
+	return true;
 }
 
-static struct k_poll_signal conn_change =
-		K_POLL_SIGNAL_INITIALIZER(conn_change);
-
-static void conn_cleanup(struct bt_conn *conn)
+void bt_conn_data_ready(struct bt_conn *conn)
 {
-	struct net_buf *buf;
-
-	/* Give back any allocated buffers */
-	while ((buf = net_buf_get(&conn->tx_queue, K_NO_WAIT))) {
-		struct bt_conn_tx *tx = tx_data(buf)->tx;
+	LOG_DBG("DR");
 
-		tx_data(buf)->tx = NULL;
-
-		/* destroy the buffer */
-		net_buf_unref(buf);
-
-		/* destroy the tx context (and any associated meta-data) */
-		if (tx) {
-			conn_tx_destroy(conn, tx);
-		}
+	/* The TX processor will call the `pull_cb` to get the buf */
+	if (!atomic_set(&conn->_conn_ready_lock, 1)) {
+		sys_slist_append(&bt_dev.le.conn_ready,
+				 &conn->_conn_ready);
+		LOG_DBG("raised");
+	} else {
+		LOG_DBG("already in list");
 	}
 
-	__ASSERT(sys_slist_is_empty(&conn->tx_pending), "Pending TX packets");
-	__ASSERT_NO_MSG(conn->pending_no_cb == 0);
-
-	bt_conn_reset_rx_state(conn);
-
-	k_work_reschedule(&conn->deferred_work, K_NO_WAIT);
+	/* Kick the TX processor */
+	bt_tx_irq_raise();
 }
 
-static void conn_destroy(struct bt_conn *conn, void *data)
+static bool cannot_send_to_controller(struct bt_conn *conn)
 {
-	if (conn->state == BT_CONN_CONNECTED ||
-	    conn->state == BT_CONN_DISCONNECTING) {
-		bt_conn_set_state(conn, BT_CONN_DISCONNECT_COMPLETE);
-	}
-
-	bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+	return k_sem_count_get(bt_conn_get_pkts(conn)) == 0;
 }
 
-void bt_conn_cleanup_all(void)
+static bool dont_have_methods(struct bt_conn *conn)
```

Comment:
Isn't it more common to name boolean functions after the `true` case? 
i.e. `cannot_send_to_controller` -> `can_send_to_controller` and `dont_have_methods` -> `have_methods`

Reply:
Ditto for `dont_have_tx_context`

Reply:
Yes. But what if they're only ever used with a `!` in front? 
IMO `if (!have_resource)` is less readable at first glance than `if (dont_have_resource)`.

Reply:
That's a fair point. You could (with good reason) argue that the function should be named after the most common scenario.